### PR TITLE
feat: add `SliceExec`

### PR DESCRIPTION
**Pull request by iajoiner**
Please be sure to look over the pull request guidelines here: https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/CONTRIBUTING.md#submit-pr.

# Please go through the following checklist
- [x] The PR title and commit messages adhere to guidelines here: https://github.com/spaceandtimelabs/sxt-proof-of-sql/blob/main/CONTRIBUTING.md. In particular `!` is used if and only if at least one breaking change has been introduced.
- [x] I have run the ci check script with `source scripts/run_ci_checks.sh`.
- The following upstream PRs have been merged:
  - [x] #381
  - [x] #401
  - [x] #404

# Rationale for this change
This PR replaces #121 and is designed to test whether our `ProofPlan`s are truly composable now.

# What changes are included in this PR?
- add `SliceExec`.

# Are these changes tested?

Yes.

### CODE REVIEW

**Review by JayWhite2357 — CHANGES_REQUESTED** [review]

**Comment by JayWhite2357** [comment]
For my reference: the review was over https://github.com/spaceandtimelabs/sxt-proof-of-sql/commit/86abb7c074f94c7af21c78d1b0af5bcade4cad6c.

File: crates/proof-of-sql/src/sql/proof_plans/slice_exec_test.rs
[the remark this answers is not in the document]
Comment (iajoiner) [reply crates/proof-of-sql/src/sql/proof_plans/slice_exec_test.rs:1]: Sure. In this case we need to error out though due to the comment you made above.

File: crates/proof-of-sql/src/sql/proof_plans/slice_exec.rs [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+}
+
+/// Get the boolean slice selection from the number of rows, skip and fetch
+fn get_slice_select(num_rows: usize, skip: usize, fetch: Option<usize>) -> Vec<bool> {
+    if let Some(fetch) = fetch {
+        let end = skip + fetch;
+        (0..num_rows).map(|i| i >= skip && i < end).collect()
+    } else {
+        (0..num_rows).map(|i| i >= skip).collect()
+    }
```
[the remark this answers is not in the document]
Comment (iajoiner) [reply crates/proof-of-sql/src/sql/proof_plans/slice_exec.rs:40]: This is neat. Thanks!

File: crates/proof-of-sql/src/sql/proof_plans/slice_exec.rs [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+/// `ProofPlan` for queries of the form
+/// ```ignore
+///     <ProofPlan> LIMIT <fetch> [OFFSET <skip>]
+/// ```
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct SliceExec {
+    pub(super) input: Box<DynProofPlan>,
+    pub(super) skip: usize,
+    pub(super) fetch: Option<usize>,
```
[the remark this answers is not in the document]
Comment (iajoiner) [reply crates/proof-of-sql/src/sql/proof_plans/slice_exec.rs:30]: I think the issue here is the ordering. `LIMIT` corresponds to `fetch` and `OFFSET` corresponds to `skip`.

**Review by JayWhite2357 — DISMISSED** [review]
FYI: I rebased after #404.
Approved.
I turned of auto-merge so you can address the NITs before merging.

**Review by JayWhite2357 — APPROVED** [review]

File: crates/proof-of-sql/src/sql/proof_plans/mod.rs [hunk 3]
```diff
@@ -16,7 +16,7 @@ mod projection_exec_test;
 pub(crate) mod test_utility;
 
 mod filter_exec;
-pub(crate) use filter_exec::FilterExec;
+pub(super) use filter_exec::FilterExec;
```
[the remark this answers is not in the document]
Comment (iajoiner) [reply crates/proof-of-sql/src/sql/proof_plans/mod.rs]: Sorry that was a mistake

**Pull request merged**